pkg/scanners/scannerv4/convert.go

```diff
@@ -0,0 +1,143 @@
+package scannerv4
+
+import (
+	"strings"
+
+	gogotypes "github.com/gogo/protobuf/types"
+	v4 "github.com/stackrox/rox/generated/internalapi/scanner/v4"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/clair"
+	"github.com/stackrox/rox/pkg/set"
+)
+
+func imageScan(metadata *storage.ImageMetadata, report *v4.VulnerabilityReport) *storage.ImageScan {
+	scan := &storage.ImageScan{
+		// ScannerVersion: ,
+		ScanTime:        gogotypes.TimestampNow(),
+		OperatingSystem: os(report),
+		Components:      components(metadata, report),
+	}
+
+	if scan.GetOperatingSystem() == "unknown" {
+		scan.Notes = append(scan.Notes, storage.ImageScan_OS_UNAVAILABLE)
+	}
+
+	return scan
+}
+
+func components(metadata *storage.ImageMetadata, report *v4.VulnerabilityReport) []*storage.EmbeddedImageScanComponent {
+	layerSHAToIndex := clair.BuildSHAToIndexMap(metadata)
+
+	components := make([]*storage.EmbeddedImageScanComponent, 0, len(report.GetPackageVulnerabilities()))
+	for _, pkg := range report.GetContents().GetPackages() {
+		id := pkg.GetId()
+		vulnIDs := report.GetPackageVulnerabilities()[id].GetValues()
+		component := &storage.EmbeddedImageScanComponent{
+			Name:          pkg.GetName(),
+			Version:       pkg.GetVersion(),
+			Vulns:         vulnerabilities(report.GetVulnerabilities(), vulnIDs),
+			Location:      pkg.GetPackageDb(),
+			HasLayerIndex: layerIndex(layerSHAToIndex, report, id),
+		}
+
+		components = append(components, component)
+	}
+
+	return components
+}
+
+func layerIndex(layerSHAToIndex map[string]int32, report *v4.VulnerabilityReport, pkgID string) *storage.EmbeddedImageScanComponent_LayerIndex {
+
+	envList := report.GetContents().GetEnvironments()[pkgID]
+	if len(envList.GetEnvironments()) > 0 {
+		env := envList.GetEnvironments()[0]
+
+		if val, ok := layerSHAToIndex[env.GetIntroducedIn()]; ok {
+			return &storage.EmbeddedImageScanComponent_LayerIndex{
+				LayerIndex: val,
+			}
+		}
+	}
+
+	return nil
+}
+
+func vulnerabilities(vulnerabilities map[string]*v4.VulnerabilityReport_Vulnerability, ids []string) []*storage.EmbeddedVulnerability {
+	if len(vulnerabilities) == 0 {
+		return nil
+	}
+
+	vulns := make([]*storage.EmbeddedVulnerability, 0, len(ids))
+	uniqueVulns := set.NewStringSet()
+	for _, id := range ids {
+		ccVuln := vulnerabilities[id]
```

Comment:
should we sanity check this `ccVuln != nil`? `ccVuln, ok := vulnerabilities[id]`